channel.go

```diff
@@ -171,19 +171,26 @@ func (ch *Channel) open(ctx context.Context) error {
 func (ch *Channel) call(ctx context.Context, req message, res ...message) error {
 	select {
 	case <-ctx.Done():
-		return ctx.Err()
+		return ErrCanceled
 	default:
 	}
 
+	if req.wait() {
```

Comment:
The reply from a cancelled request may be in-flight and not reside on the `rpc` at this point and it will not drain.

Take the scenario all with different contexts:

```
client send(a)
client send(b)
client wait(a) - cancelled - returned
server reply(a)
client wait(b) - recv reply(a)
```

Reply:
i don't think this use case is possible. Every RPC call is surounded by a mutex.

Therefore that could end up as:
```
client send(a)
client wait(a) - cancelled - returned
server reply(a) - message dropped
client send(b)
client wait(b) 
server reply(b) 
recv reply(b) 
```

Or (the case i exposed here https://github.com/streadway/amqp/pull/389#issuecomment-467143264)

```
client send(a)
client wait(a) - cancelled - returned
client send(b)
client wait(b) 
server reply(a) - message dropped
recv reply(a)  
```
And it causes issues.

But i can not figure out any way to handle this case with the current protocol. IMHO the safest and most deterministic way of dealing with this situation is to close the channel in case of cancelation.

Reply:
I still think the most obvious behavior is not to close the channel when a context is `Done()`. I believe context done is closer to a late `no-wait` rather than a client error.

This is a matter of sequencing the consumption of the replies in the same order as the requests were written. It looks like communication between the writer goroutine (application) and reader goroutine (dispatch) is how the sequencing would be implemented.